### PR TITLE
JPO-ODE 5.0.2 Support

### PR DESCRIPTION
**Pull request by drewjj**
## Description
Adds support for the latest jpo-ode version (5.0.2) which introduces breaking changes to the output of the OdeBsmJson topic to conform to the J2735 standard better.

Main Updates:
- Updated the PPM logic to traverse the BSM messages according to the newest OdeBsmJson output.
- Added `metadata.originIp` to the always redacted fields defined in the PPM.
- Resolved a bug with redactions involving fields in objects that are elements of an array. This previously would only redact the first element of an array's specified field, but now it will redact the field from all elements of the array.
- Removed any mentioning of unique bitstring handling since this logic is no longer used or relevant to the latest OdeBsmJson output. A few unit tests have also been removed that specifically targeted bitstring testing.
- Removed old WYDOT test BSM data since it was no longer correct and functional with the new PPM. The I-80 edge file has been retained in case new data is added in the future that is compatible with it.
- Added CO-Roadways edge file and example OdeBsmJson JSON.
- Updated the fieldsToRedact.txt with the new field paths and updated the ppmBsm.properties to correspond with the CO-Motorways edge file.
- Updated integration testing scripts and the unit tests to work with the latest functionality of the PPM and the format of the OdeBsmJson output.

Minor Updates:
- I noticed the latest version of develop had some inconsistencies from main which was up to date with the USDOT version of the jpo-cvdp. I merged the latest up stream changes with the updates from develop into my feature branch to fix this inconsistency. This is where the Azure pipeline changes come from. I did not add these or change them.
- Minor CI formatting change and if statement for the Dockerhub job to ensure it only runs in the USDOT repository.

## Testing
Tested with the following methods:
- Unit tests following the procedure defined in the [testing .md](https://github.com/CDOT-CV/jpo-cvdp/blob/fix/bsm-format-update/docs/testing.md#unit-testing) documentation.
- Integration tests by running the `./do_kafka_test.sh` script with Docker running.
- Running the PPM standalone in Docker without the wurstmeister zookeeper and kafka while also running a standard docker-compose deployment of the jpo-ode 5.0.2 repository. I then run the UDP send scripts to send payloads to the jpo-ode and ensure it makes it all the way to the FilteredOdeBsmJson topic with the expected redactions made.

Here are the .env values I used for Docker:
```
# The host machine that the PPM is running on.
DOCKER_HOST_IP=

# An external directory that contains the necessary edges and properties files.
# I created this directory in my clone directory, it is also already in the .gitignore
# Make sure this includes your edge file (CO-Motorways.edges), your properties file (ppmBsm.properties), 
# and your redaction file (fieldsToRedact.txt)
DOCKER_SHARED_VOLUME=~/jpo-cvdp/ppm_data

# The config file to use. This file must be in the shared volume.
PPM_CONFIG_FILE=ppmBsm.properties

# The path to the fieldsToRedact.txt file needed by the RedactionPropertiesManager class. This file must be in the shared volume.
REDACTION_PROPERTIES_PATH=/ppm_data/fieldsToRedact.txt

# Logging related flags for file/console logging & log level
PPM_LOG_TO_FILE=false
PPM_LOG_TO_CONSOLE=true
# Valid values: TRACE, DEBUG, INFO, WARN, ERROR, CRITICAL
PPM_LOG_LEVEL=WARN

# When set to true, the RPM will print debug messages to a file. (this is separate from regular logging)
RPM_DEBUG=false

# This should be "CONFLUENT" when utilizing CC and empty otherwise.
KAFKA_TYPE=

# Key and secret for CC.
CONFLUENT_KEY=
CONFLUENT_SECRET=
```
Example payload to send to the jpo-ode that contains all of the fields in the `fieldsToRedact.txt`:

`001480955F93D0D5116EE8A6AD7139165E9182A70581E6772631B3461EDC56B89DB8AEFD2208AE12F90151E10037FFF2B86414E7F9925EB4A3C08281CDE9E67DCCEC75DFF153F9922C850EA042702EEA2F72F4F751394C54B0488322BA2BA4051BA5E702510E8024DDF9171DE8EE3FFAACC4B3F724FD3F6EED2DEB566AF0082186C8F6FF62A10865D1D84FD5532A3AED757718E1C2CB5D0340`

### CODE REVIEW

File: src/general-redaction/rapidjsonRedactor.cpp [hunk 1]
```diff
@@ -112,15 +91,16 @@ bool RapidjsonRedactor::redactMemberByPath(rapidjson::Value &value, std::string
         }
     }
     else if (value.IsArray()) {
+        bool result = false;
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/general-redaction/rapidjsonRedactor.cpp:94]: I think Copilot is mistaken here. I have attached the full code block. While the `result` variable can be set to true inside the loop, it is predefined as false which makes this whole comment not relevant since it is always going to at least return false but true if it does happen. Maybe the uncertainty of whether it will ever change to true is what triggered this concern, but this is as intended.

```
else if (value.IsArray()) {
    bool result = false;
    for (auto &m : value.GetArray()) {
        std::string type = kTypeNames[m.GetType()];
        if (type == "Object" || type == "Array") {
            if (redactMemberByPath(m, path)) {
                result = true;
            }
        }
    }
    return result;
}
```

File: src/bsmHandler.cpp [hunk 2]
```diff
@@ -345,20 +367,21 @@ bool BSMHandler::process( const std::string& message_json ) {
 void BSMHandler::handleGeneralRedaction(rapidjson::Document& document) {
     if (is_active<kGeneralRedactFlag>()) {
         for (std::string memberPath : rpm.getFields()) {
+            logger_->info("Attempting redaction for path: '" + memberPath + "'");
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp]: Committed suggestion. I added this log for testing and had intended to remove it but it was very useful so debug is fitting.

File: src/bsmHandler.cpp [hunk 3]
```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
...
 
             return false;
         }
 
-        if (!position["latitude"].IsDouble() || !position["longitude"].IsDouble()) {
-            result_ = ResultStatus::OTHER;
-
-            return false;
+        // Only set latitude if it's not the J2735 "unavailable" value
+        if (core_data["lat"].GetInt() != 900000001) {
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp]: Addressed by adding the constant.

File: src/general-redaction/rapidjsonRedactor.cpp [hunk 4]
```diff
@@ -77,7 +51,12 @@ bool RapidjsonRedactor::redactMemberByPath(rapidjson::Value &value, std::string
                         return true;
                     }
                     else if (type == "String" && target == "transmission") {
-                        value["transmission"] = "UNAVAILABLE";
+                        value["transmission"] = "unavailable";
```
Comment (drewjj) [inline src/general-redaction/rapidjsonRedactor.cpp:54]: This was changed to lower case to match how the OdeBsmJson output writes it. This is how the J2735 defines it as well.

File: src/bsmHandler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp]: Nevermind this. I have removed this line altogether. This logger doesn't support debug and I think it just makes sense to remove this logging line since it was intended to be removed anyways.

File: data/I_80_test.json [hunk 5]
```diff
@@ -1,11 +0,0 @@
...": 71.7}, {"elevationOffset": 204.7, "latOffset": 0.0001313, "lonOffset": 6.06e-05, "timeOffset": 80.2}, {"elevationOffset": 204.7, "latOffset": 0.0001535, "lonOffset": 7.48e-05, "timeOffset": 92.9}, {"elevationOffset": 204.7, "latOffset": 0.000165, "lonOffset": 0.0001171, "timeOffset": 126.4}, {"elevationOffset": 204.7, "latOffset": 0.0001363, "lonOffset": 0.0001, "timeOffset": 136.1}, {"elevationOffset": 204.7, "latOffset": 0.0001196, "lonOffset": 0.0001028, "timeOffset": 141.9}, {"elevationOffset": 204.7, "latOffset": 0.0001111, "lonOffset": 9.8e-05, "timeOffset": 143.0}, {"elevationOffset": 204.7, "latOffset": 0.0131071, "lonOffset": 9.5e-06, "timeOffset": 348.0}, {"elevationOffset": 12.6, "latOffset": 0.0131071, "lonOffset": 0.0131071, "timeOffset": 437.8}, {"elevationOffset": 11.5, "latOffset": 0.0131071, "lonOffset": 0.0131071, "timeOffset": 468.2}, {"elevationOffset": 10.0, "latOffset": 0.0131071, "lonOffset": 0.0131071, "timeOffset": 474.0}, {"elevationOffset": 10.2, "latOffset": 0.0131071, "lonOffset": 7e-06, "timeOffset": 481.3}]}, "pathPrediction": {"confidence": 0.0, "radiusOfCurve": 0.0}}}]}, "dataType": "us.dot.its.jpo.ode.plugin.j2735.J2735Bsm"}, "schemaVersion": 3}
```
Comment (payneBrandon) [inline data/I_80_test.json:1]: question: was this file unused? or invalid format now?

File: data/CO-Motorways.edges
[the remark this answers is not in the document]
Comment (payneBrandon) [reply data/CO-Motorways.edges:1]: question: I believe we'd left out CDOT specific data at the request of Dan previously - she was concerned with adding project specific data to the general repo. Is this generalized enough to get around that?

File: .azure/azure-pipelines.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# Azure pipeline used to trigger builds of the JPO CVDP project
+# This pipeline is primarily used for CI in the CDOT-CV fork
+
+trigger:
+  branches:
+    include:
+      - develop
```
Comment (payneBrandon) [inline .azure/azure-pipelines.yml:7]: question: I see that in other repos we also trigger off of the dev/develop branch, why have we chosen that over master or some "release/*" nomenclature?

File: .github/workflows/dockerhub.yml [hunk 7]
```diff
@@ -8,6 +8,7 @@ on:
 
 jobs:
   publish:
+    if: github.event.pull_request.base.repo.owner.login == 'usdot-jpo-ode'
```
Comment (payneBrandon) [inline .github/workflows/dockerhub.yml:11]: similar suggestion here, there's a potential exception on push that isn't part of a PR

File: config/fieldsToRedact.txt [hunk 8]
```diff
@@ -1,38 +1,22 @@
...
+payload.data.value.BasicSafetyMessage.coreData.angle
+payload.data.value.BasicSafetyMessage.coreData.transmission
+payload.data.value.BasicSafetyMessage.coreData.brakes.wheelBrakes
+payload.data.value.BasicSafetyMessage.coreData.brakes.traction
+payload.data.value.BasicSafetyMessage.coreData.brakes.abs
+payload.data.value.BasicSafetyMessage.coreData.brakes.scs
+payload.data.value.BasicSafetyMessage.coreData.brakes.brakeBoost
+payload.data.value.BasicSafetyMessage.coreData.brakes.auxBrakes
+payload.data.value.BasicSafetyMessage.partII.partII-Value.VehicleSafetyExtensions.events
+payload.data.value.BasicSafetyMessage.partII.partII-Value.VehicleSafetyExtensions.lights
```
Comment (payneBrandon) [inline config/fieldsToRedact.txt:10]: question: is this just redacting everything _under_ lights rather than iterating through all the values (presumably what we did before)?

File: include/bsmHandler.hpp [hunk 9]
```diff
@@ -124,6 +124,9 @@ class BSMHandler {
         static constexpr uint32_t kIdRedactFlag       = 0x1 << 2;
         static constexpr uint32_t kSizeRedactFlag     = 0x1 << 4;
         static constexpr uint32_t kGeneralRedactFlag  = 0x1 << 8;
+        static constexpr int J2735_SPEED_UNAVAILABLE = 8191;
```
Comment (payneBrandon) [inline include/bsmHandler.hpp:127]: question: where are these values coming from? can we add a comment letting devs know this?

File: src/bsmHandler.cpp [hunk 10]
```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
...
             result_ = ResultStatus::OTHER;
 
             return false;
         }
 
-        speed = core_data["speed"].GetDouble();
-        bsm_.set_velocity(speed);
+        // Only set speed if it's not the J2735 "unavailable" value
+        if (core_data["speed"].GetInt() != J2735_SPEED_UNAVAILABLE) {
+            speed = core_data["speed"].GetInt() * 0.02;
```
Comment (payneBrandon) [inline src/bsmHandler.cpp:267]: suggestion: add comment explaining why we're multiplying by the .02 value

File: src/bsmHandler.cpp [hunk 11]
```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
...
         }
-        
-        latitude = position["latitude"].GetDouble();
-        longitude = position["longitude"].GetDouble();
 
-        bsm_.set_latitude(latitude); 
-        bsm_.set_longitude(longitude); 
+        // Only set longitude if it's not the J2735 "unavailable" value
+        if (core_data["long"].GetInt() != J2735_LONGITUDE_UNAVAILABLE) {
+            longitude = core_data["long"].GetInt() * 1e-7;
```
Comment (payneBrandon) [inline src/bsmHandler.cpp:296]: suggestion: add comment explaining the math

File: data/I_80_test.json (hunk 5 above)
Comment (drewjj) [reply data/I_80_test.json:1]: Invalid. No way of updating it without the original ASN.1

File: .azure/azure-pipelines.yml (hunk 6 above)
Comment (drewjj) [reply .azure/azure-pipelines.yml:7]: I did not add this so I don't have insight. I can make it consistent.

File: config/bsm-test/c1.properties [hunk 12]
```diff
@@ -17,10 +17,10 @@ privacy.filter.geofence.mapfile=/ppm_data/road_file.csv
 privacy.redaction.size=OFF
 
 # geographic bounds that can further restrict the map information.
-privacy.filter.geofence.sw.lat=40.970819
-privacy.filter.geofence.sw.lon=-111.090454
-privacy.filter.geofence.ne.lat=42.136908
-privacy.filter.geofence.ne.lon=-103.793068
+privacy.filter.geofence.sw.lat=37.002
```
[the remark this answers is not in the document]
Comment (drewjj) [reply config/bsm-test/c1.properties:20]: They are. I didn't have WYDOT data so I went with what I had to work with.

File: include/bsmHandler.hpp (hunk 9 above)
Comment (drewjj) [reply include/bsmHandler.hpp:127]: J2735 and the constant's name is fairly indicative of that to be honest. I can add a comment I suppose.

File: src/bsmHandler.cpp (hunk 10 above)
Comment (drewjj) [reply src/bsmHandler.cpp:267]: This technically has the comment that explains the value used as well.

File: .github/workflows/dockerhub.yml (hunk 7 above)
Comment (drewjj) [reply .github/workflows/dockerhub.yml]: Addressed

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -1,98 +1,100 @@
...
+          sonar.exclusions=**/*.java
+          sonar.coverage.exclusions=**/*.java
+          # Set Git as SCM sensor
+          sonar.scm.disabled=true
+          #sonar.scm.enabled=false
+          sonar.scm.provider=git
+          sonar.sourceEncoding=UTF-8
+          EOF
+      - name: Run SonarScanner
+        if: github.event.pull_request.base.repo.owner.login == 'usdot-jpo-ode'
```
[the remark this answers is not in the document]
Comment (drewjj) [reply .github/workflows/ci.yml]: Committed suggestion.

File: src/bsmHandler.cpp (hunk 10 above)
Comment (drewjj) [reply src/bsmHandler.cpp:268]: Addressed.

File: src/bsmHandler.cpp [hunk 14]
```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
...
             return false;
         }
 
-        if (!position["latitude"].IsDouble() || !position["longitude"].IsDouble()) {
-            result_ = ResultStatus::OTHER;
-
-            return false;
+        // Only set latitude if it's not the J2735 "unavailable" value
+        if (core_data["lat"].GetInt() != J2735_LATITUDE_UNAVAILABLE) {
+            latitude = core_data["lat"].GetInt() * 1e-7;
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp:292]: Addressed.

File: src/bsmHandler.cpp (hunk 11 above)
Comment (drewjj) [reply src/bsmHandler.cpp:299]: Addressed.

File: config/fieldsToRedact.txt [hunk 15]
```diff
@@ -1,38 +1,22 @@
...
+payload.data.value.BasicSafetyMessage.partII.partII-Value.VehicleSafetyExtensions.pathHistory.crumbData.timeOffset
+payload.data.value.BasicSafetyMessage.partII.partII-Value.SupplementalVehicleExtensions.doNotUse1.solarRadiation
```
[the remark this answers is not in the document]
Comment (drewjj) [reply config/fieldsToRedact.txt:12]: I do think for testing purposes it does kind of provide some benefit to have them individually defined. I also wanted to only change things that needed to change in order to preserve as much functionality as possible.

File: src/general-redaction/rapidjsonRedactor.cpp (hunk 4 above)
Comment (drewjj) [reply src/general-redaction/rapidjsonRedactor.cpp:56]: Addressed.

File: config/fieldsToRedact.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (drewjj) [reply config/fieldsToRedact.txt:12]: I will give a try! I think we need to keep it down to the deepest child node in the unit tests because of how it tries to identify edge nodes but for this general redact file I think it would work just fine.

[the remark this answers is not in the document]
Comment (drewjj) [reply config/fieldsToRedact.txt:12]: I have consolidated the doNotUse2 and doNotUse4 cases since doNotUse1 is not technically treated the same way and can have individual fields redacted without having the whole object removed.

File: src/bsmHandler.cpp [hunk 16]
```diff
@@ -222,58 +226,79 @@ bool BSMHandler::process( const std::string& message_json ) {
...
-        bsm_.set_velocity(speed);
+        // Only set speed if it's not the J2735 "unavailable" value
+        if (core_data["speed"].GetInt() != J2735_SPEED_UNAVAILABLE) {
+            // J2735 defined INTEGER (0..8190) -- Units of 0.02 m/s
+            speed = core_data["speed"].GetInt() * 0.02;
+            bsm_.set_velocity(speed);
+        }
 
         if (is_active<kVelocityFilterFlag>() && vf_.suppress(speed)) {
             result_ = ResultStatus::SPEED;
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp:273]: Yes you are right. I have added the return false when the speed suppression is true.

File: src/bsmHandler.cpp [hunk 17]
```diff
@@ -222,58 +226,81 @@ bool BSMHandler::process( const std::string& message_json ) {
...
+            latitude = core_data["lat"].GetInt() * 1e-7;
+            bsm_.set_latitude(latitude);
+        }
+
+        // Only set longitude if it's not the J2735 "unavailable" value
+        if (core_data["long"].GetInt() != J2735_LONGITUDE_UNAVAILABLE) {
+            // J2735 defined INTEGER (-1799999999..1800000000) --  Units of 1/10 microdegree
+            longitude = core_data["long"].GetInt() * 1e-7;
+            bsm_.set_longitude(longitude);
+        }
```
Comment (payneBrandon) [inline src/bsmHandler.cpp:304]: suggestion: similar suggestion here as above - I believe we can set this to return false in this check. We should also look at moving all the checks above any of our setters to enhance performance

File: src/bsmHandler.cpp [hunk 18]
```diff
@@ -222,58 +226,79 @@ bool BSMHandler::process( const std::string& message_json ) {
...
-        if (!position["latitude"].IsDouble() || !position["longitude"].IsDouble()) {
-            result_ = ResultStatus::OTHER;
-
-            return false;
+        // Only set latitude if it's not the J2735 "unavailable" value
+        if (core_data["lat"].GetInt() != J2735_LATITUDE_UNAVAILABLE) {
+            // J2735 defined INTEGER (-900000000..900000000) --  Units of 1/10 microdegree
+            latitude = core_data["lat"].GetInt() * 1e-7;
+            bsm_.set_latitude(latitude);
         }
```
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp:296]: I am not sure I have a strong opinion on this. I chose to leave it as the ppm's defaults which is actually -1 speed and (90.0, 180.0) latitude and longitude values.  This automatically get redacted whenever the speed is not available and it is highly likely the position data will be as well.

I could see automatically filtering the message out and short-cutting the logic right there or allowing it to be included. I don't think we should add logic around the redaction of speed and positioning to support a null case or something like that.

File: src/bsmHandler.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (drewjj) [reply src/bsmHandler.cpp]: Exactly.

File: src/bsmHandler.cpp (hunk 17 above)
Comment (drewjj) [reply src/bsmHandler.cpp:316]: Okay this broke the unit tests. I will have to see how to go about doing this correctly.

File: src/bsmHandler.cpp [hunk 19]
```diff
@@ -222,61 +226,85 @@ bool BSMHandler::process( const std::string& message_json ) {
...
 
             return false;
         }
 
-        speed = core_data["speed"].GetDouble();
-        bsm_.set_velocity(speed);
+        // Only set speed if it's not the J2735 "unavailable" value
+        if (core_data["speed"].GetInt() != J2735_SPEED_UNAVAILABLE) {
+            // J2735 defined INTEGER (0..8190) -- Units of 0.02 m/s
+            speed = core_data["speed"].GetInt() * 0.02;
```
Comment (mcook42) [inline src/bsmHandler.cpp:268]: question(non-blocking): would it make sense to call `bsm_.set_velocity(speed)` here instead of on line 333 so the calculation of speed is done in the same place as where we set speed? Or is there a functional reason we wait until line 333 to set velocity?

**Review by mcook42 — APPROVED** [review]

File: data/CO-Motorways.edges
[the remark this answers is not in the document]
Comment (payneBrandon) [reply data/CO-Motorways.edges:1]: I agree, the additional data is probably a good thing and indicates additional potential boundaries for other stated to reference when implementing. We'll see if they have any qualms with it on merge.
